common/emplace_result.h

```diff
@@ -0,0 +1,64 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_EMPLACE_RESULT_H_
+#define CARBON_COMMON_EMPLACE_RESULT_H_
+
+#include <type_traits>
+#include <utility>
+
+namespace Carbon {
+
+// A utility to use when calling an `emplace` function to emplace the result of
+// a function call. Expected usage is:
+//
+//   my_widget_vec.emplace_back(EmplaceResult([&] {
+//     return ConstructAWidget(...);
+//   }));
+//
+// In this example, the result of `ConstructAWidget` will be constructed
+// directly into the new element of `my_widget_vec`, without performing a copy
+// or move.
+//
+// This works by providing a conversion function on `EmplaceResult` that
+// converts to the type `DestT` being emplaced. When an `emplace` function is
+// used to construct an object of type `DestT` from a value of type
+// `EmplaceResult` that converts to `DestT`, the conversion function is used
+// directly to initialize the `DestT` object without calling a `DestT`
+// constructor, per the C++17 guaranteed copy elision rules. Similarly, the
+// result of the conversion function is initialized directly by calling
+// `make_fn`, again relying on guaranteed copy elision.
+//
+// Because the make function is called from the conversion function,
+// `EmplaceResult` should only be used in contexts where it will be used to
+// initialize a `DestT` object exactly once. This is generally true of `emplace`
+// functions. Also, because the `make_fn` callback will be called after the
+// container has made space for the new element, it should not inspect or modify
+// the container that is being emplaced into.
+template <typename MakeFnT>
+class EmplaceResult {
```

Comment:
nit: I might call this EmplaceConstruct as it's what the job of the thing is, and it's not the result of the emplacing operation. Up to you.

Reply:
I'm not tied to this particular name, but `EmplaceConstruct` seems to be missing some important information -- a reference to the fact that it's calling a function and emplacing the result of that function call. I'd also be happy with things like `EmplaceByCalling(callable)` or `EmplaceResultOf(callable)` that avoid the possibility of this name being interpreted as "(the) emplace(ment) result" instead of "emplace (the) result (of)".

Reply:
Both of those would be fine with me, yeah.

Reply:
OK, I'll go with `EmplaceByCalling` -- on further thought I think `EmplaceResultOf` sounds too much like a type trait analogous to `std::result_of`.